core/network/impl/synchronizer_impl.cpp

```diff
@@ -1687,12 +791,325 @@ namespace kagome::network {
+        }
+        case VisitAncestryResult::IGNORE_SUBTREE: {
+          break;
+        }
+      }
+    }
+  }
+
+  bool SynchronizerImpl::canFetchMore(size_t &fetching) const {
+    return fetching < max_parallel_downloads_;
+  }
+
+  bool SynchronizerImpl::validate(const BlocksRequest &request,
+                                  BlocksResponse &response) const {
+    auto &blocks = response.blocks;
+    if (blocks.empty()) {
+      return false;
+    }
+    auto need_header = has(request.fields, BlockAttribute::HEADER);
+    auto need_body = has(request.fields, BlockAttribute::BODY);
+    if (auto *hash = boost::get<BlockHash>(&request.from)) {
+      if (blocks[0].hash != *hash) {
+        return false;
+      }
+    } else {
+      auto number = boost::get<BlockNumber>(request.from);
+      if (blocks[0].header.has_value() and blocks[0].header->number != number) {
+        return false;
+      }
+    }
+    if (request.direction == Direction::ASCENDING) {
+      std::ranges::reverse(blocks);
+    }
+    for (auto &block : blocks) {
+      if (block.header.has_value() != need_header) {
+        return false;
+      }
+      if (not need_body and block.body.has_value()) {
+        return false;
+      }
+      if (need_header) {
+        calculateBlockHash(block.header.value(), *hasher_);
+        if (block.hash != block.header->hash()) {
+          return false;
+        }
+      }
+    }
+    if (need_header) {
+      std::optional<BlockInfo> expected;
+      size_t count = 0;
+      for (auto &block : blocks) {
+        auto &header = block.header.value();
+        ++count;
+        if (expected.has_value() and header.blockInfo() != expected) {
+          return false;
+        }
+        expected = header.parentInfo();
+        if (not expected.has_value() and count != blocks.size()) {
+          return false;
+        }
+      }
+    }
+    if (need_body) {
+      size_t count = 0;
+      for (auto &block : blocks) {
+        ++count;
+        if (not block.body.has_value()) {
+          if (count != blocks.size()) {
+            return false;
+          }
+          continue;
+        }
+        if (auto known = entry(known_blocks_, block.hash)) {
+          if (not checkExtrinsicRoot(known->data.header.value(), *block.body)) {
+            return false;
+          }
+        }
+      }
+    }
+    return true;
+  }
+
+  size_t SynchronizerImpl::highestAllowedBlock() const {
+    return block_tree_->bestBlock().number
+         + (sync_method_ == application::SyncMethod::Full
+                ? kMinPreloadedBlockAmount
+                : kMinPreloadedBlockAmountForFastSyncing);
+  }
+
+  bool SynchronizerImpl::isBlockAllowed(const BlockHeader &header) const {
+    auto finalized = block_tree_->getLastFinalized();
+    if (header.number <= finalized.number) {
+      return false;
+    }
+    if (header.number == finalized.number + 1
+        and header.parent_hash != finalized.hash) {
+      return false;
+    }
+    return true;
```

Comment:
Additional check is needed, base on current time and slot. All blocks with a slot greater by 1-2 than the current one should be considered malicious.

Reply:
Added check for increasing slot number.
I think that blocks shouldn't be ignored on slot difference, because they can be correct (e.g. restarting all validators, or network problems).

Reply:
I meant block with slot-number from future, disaccorded with current time more than two slot durations.